File: elixir/binary-search-tree/lib/binary_search_tree.ex
defmodule BinarySearchTree do
  @type bst_node :: %{data: any, left: bst_node | nil, right: bst_node | nil}

  @doc """
  Create a new Binary Search Tree with root's value as the given 'data'
  """
  @spec new(any) :: bst_node
  def new(data) do
    %{data: data, left: nil, right: nil}
  end

  @doc """
  Creates and inserts a node with its value as 'data' into the tree.
  """
  @spec insert(bst_node, any) :: bst_node
  def insert(tree, data) do
    if data <= tree.data do
      if tree.left == nil do
        %{tree | left: new(data)}
      else
        %{tree | left: insert(tree.left, data)}
      end
    else
      if tree.right == nil do
        %{tree | right: new(data)}
      else
        %{tree | right: insert(tree.right, data)}
      end
    end
  end

  @doc """
  Traverses the Binary Search Tree in order and returns a list of each node's data.
  """
  @spec in_order(bst_node) :: [any]
  def in_order(tree) do
    if tree == nil do
      []
    else
      in_order(tree.left) ++ [tree.data] ++ in_order(tree.right)
    end
  end
end

File: fsharp/binary-search-tree/BinarySearchTree.fs
module BinarySearchTree

type Node = 
    | Null
    | Node of Node * int * Node

let left = function
    | Null | Node(Null, _, _) -> None
    | Node(l, _, _) -> Some l

let right = function 
    | Null | Node(_, _, Null) -> None
    | Node(_, _, r) -> Some r 

let data = function
    | Null -> failwith "No data for empty tree"
    | Node(_, d, _) -> d 

let rec create = function
    | [] -> Null 
    | x::xs  -> 
        let (ys, zs) = xs |> List.partition ((>=) x)
        Node(create ys, x, create zs) 
    
let rec sortedData = function
    | Null -> []
    | Node(l, d, r) -> sortedData l @ [d] @ sortedData r